File: backend/internal/gpt/gpt.py
import json
import logging
import os
from fastapi import FastAPI, Request, WebSocket, WebSocketDisconnect
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel
from openai import OpenAI
from dotenv import load_dotenv
from .prompts import get_hint_prompt, get_question_prompt


load_dotenv()
if not os.getenv("GPT_API_KEY"):
    raise ValueError("GPT_API_KEY is missing in environment variables.")
openai_client = OpenAI(
    base_url="https://models.github.ai/inference",
    api_key=os.getenv("GPT_API_KEY"),
)

app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:5173"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

class HintRequest(BaseModel):
    code: str
    hintType: str

class QuestionRequest(BaseModel):
    difficulty: str
    company: str
    topic: str

class Message(BaseModel):
    type: str
    code: str
    hintType: str


# GPT generation logic
async def get_gpt_response(code: str, hint_type: str) -> str:
    prompt = get_hint_prompt(code, hint_type)
    response = openai_client.chat.completions.create(
        model="openai/gpt-4o",
        messages=[
            {"role": "system", "content": "You are a helpful interviewer during a technical Software Engineering interview."},
            {"role": "user", "content": prompt}
        ],
        max_tokens=500,
        temperature=0.7,
    )

    return response.choices[0].message.content.strip()

async def generate_question(difficulty: str, company: str, topic: str) -> str:
    prompt = get_question_prompt(difficulty, company, topic)
    response = openai_client.chat.completions.create(
        model="openai/gpt-4o",
        messages=[
            {
                "role": "system",
                "content": """You are a professional technical interviewer. 
                Return only the coding interview question — do not include explanations, hints, or solutions. 
                Follow this exact format:

                1. A clear description of the problem. Each sentence is on a new line.
                2. Mention any time/space complexity constraints if relevant.
                3. Include exactly one example, labeled as: Example 1:
                4. Do not include any headers like 'Question:' or 'Problem:', or 'Title:'.
                5. Do not add greetings or prefaces like 'Sure, here is...'.

                Format all content as plain text with line breaks. Avoid Markdown or numbered lists."""
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        max_tokens=500,
        temperature=0.7,
    )

    return response.choices[0].message.content.strip()
    

# REST API endpoint
@app.post("/api/hint")
async def hint_handler(req: HintRequest):
    try:
        hint = await get_gpt_response(req.code, req.hintType)
        return {"hint": hint}
    except Exception as e:
        logging.error("Hint generation failed", exc_info=True)
        return {"error": "Failed to generate hint"}
    
@app.post("/api/generate")
async def generate_handler(req: QuestionRequest):
    try:
        question = await generate_question(req.difficulty, req.company, req.topic)
        return {"question": question}
    except Exception as e:
        logging.error("Question gneration failed", exc_info=True)
        return {"error": "Failed to generate question"}

File: backend/internal/gpt/gpt.go
package websocket

import (
	"context"
	"encoding/json"
	"fmt"
	"interviewlab-backend/internal/websocket"
	"log"
	"net/http"

	"github.com/gin-gonic/gin"
	openai "github.com/sashabaranov/go-openai"
)

type Message struct {
	Type     string `json:"type"`
	Code     string `json:"code"`
	HintType string `json:"hintType"`
}

type HintRequest struct {
	Code     string `json:"code"`
	HintType string `json:"hintType"`
}

func GetGPTResponse(client *openai.Client, code string, hintType string) (string, error) {
	var prompt string

	switch hintType {
	case "weak":
		prompt = fmt.Sprintf("I'm working on this code:\n\n%s\n\nIf there is nothing, say No code shown. Tell me if I am on the right track.Give me a vague or small hint or suggestion to improve it in one paragraph. Dont give answer and Dont say anything acknowledging me or the prompt like I am happy to help or Ok, just say the advice of logic or syntax errors. DONT FOCUS ON TIME COMPLEXITY ADVICE AT ALL", code)
	case "strong":
		prompt = fmt.Sprintf("Here is my code:\n\n%s\n\nIf there is nothing say no code shown. Tell me if I am on the right track.Please give me a very detailed or strong hint and guidance in the right direction to help me improve it in one paragraph. Do not give me the full answer, just advice to lead me in the right direction. Dont say anything acknowledging me or the prompt like I am happy to help or Ok. Just say the advice", code)
	}

	resp, err := client.CreateChatCompletion(context.Background(), openai.ChatCompletionRequest{
		Model: "openai/gpt-4o",
		Messages: []openai.ChatCompletionMessage{
			{Role: openai.ChatMessageRoleSystem, Content: "You are a helpful interviewer."},
			{Role: openai.ChatMessageRoleUser, Content: prompt},
		},
		MaxTokens:   500,
		Temperature: 0.7,
	})
	if err != nil {
		return "", err
	}

	return resp.Choices[0].Message.Content, nil
}

func HintHandler(client *openai.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Header("Access-Control-Allow-Origin", "http://localhost:5173")
		c.Header("Access-Control-Allow-Methods", "POST, OPTIONS")
		c.Header("Access-Control-Allow-Headers", "Content-Type")

		if c.Request.Method == http.MethodOptions {
			c.Status(http.StatusOK)
			return
		}

		var req HintRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
			return
		}

		hint, err := GetGPTResponse(client, req.Code, req.HintType)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate hint"})
			return
		}

		c.JSON(http.StatusOK, gin.H{"hint": hint})
	}
}

func WebSocketHandler(client *openai.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		conn, err := websocket.Upgrade(c)
		if err != nil {
			log.Println("WebSocket upgrade failed:", err)
			return
		}
		defer conn.Close()

		for {
			_, msgData, err := conn.ReadMessage()
			if err != nil {
				log.Println("WebSocket read error:", err)
				break
			}

			var msg Message
			if err := json.Unmarshal(msgData, &msg); err != nil {
				log.Println("WebSocket JSON unmarshal error:", err)
				continue
			}

			if msg.Type == "hint" {
				hint, err := GetGPTResponse(client, msg.Code, msg.HintType)
				if err != nil {
					log.Println("GPT generation error:", err)
					conn.WriteJSON(map[string]string{"error": "GPT failed to generate a response"})
					continue
				}

				conn.WriteJSON(map[string]string{
					"type": "hint",
					"hint": hint,
				})
			}
		}
	}
}